README.md
<h1 align="center">Validator</h1>

<p align="center">Lightweight Validation Like Laravel</p>

<p align="center">
<a href="https://travis-ci.org/RunnerLee/validator"><img src="https://travis-ci.org/RunnerLee/validator.svg?branch=master" /></a>
<a href="https://scrutinizer-ci.com/g/RunnerLee/validator/?branch=master"><img src="https://scrutinizer-ci.com/g/RunnerLee/validator/badges/coverage.png?b=master" title="Scrutinizer Code Coverage"></a>
<a href="https://scrutinizer-ci.com/g/RunnerLee/validator/?branch=master"><img src="https://scrutinizer-ci.com/g/RunnerLee/validator/badges/quality-score.png?b=master" title="Scrutinizer Code Quality"></a>
<a href="https://styleci.io/repos/82630238"><img src="https://styleci.io/repos/82630238/shield?branch=master" alt="StyleCI"></a>
<a href="https://github.com/RunnerLee/validator"><img src="https://poser.pugx.org/runner/validator/v/stable" /></a>
<a href="http://www.php.net/"><img src="https://img.shields.io/badge/php-%3E%3D5.6-8892BF.svg" /></a>
<a href="https://github.com/RunnerLee/validator"><img src="https://poser.pugx.org/runner/validator/license" /></a>
</p>

## 使用
```php
<?php
use Runner\Validator\Validator;

// 增加规则扩展
Validator::addExtension(
    'channel',
    function ($field, $value, array $parameters = []) {
        return false !== array_search($value, ['google', 'bing']);
    },
    false
);

$data = [
    'blog' => 'https://github.com/RunnerLee',
];
$rule = [
    'blog' => 'required|url',
];
$validator = new Validator($data, $rule);

var_dump($validator->validate()); // output: true

// 获取错误消息
$validator->messages();
```

## 内置验证规则
* `accept`：字段的取值范围必须是：yes、on、'1'、1、true 中的一个
* `numeric`：字段的值必须是数字（包括整形和浮点数）
* `integer`：字段的值必须是整形
* `float`：字段的值必须是浮点型
* `size`：字段的值必须为指定长度
* `url`：字段的值必须是合法的 url 类型
* `boolean`：字段的值必须是布尔类型
* `confirm`：字段的值必须与指定值相等
* `date`：字段的值必须是一个合法的日期
* `email`：字段的值必须是一个合法的 email
* `required`：必须包含该字段
* `requiredWith`：如果指定的字段中的任意一个有值且不为空，则此字段为必填
* `requiredWithout`：如果缺少任意一个指定的字段，则此字段为必填
* `requiredIf`：如果指定的其它字段（anotherfield）等于任何一个 value 时，此字段为必填
* `array`：字段的值必须是数组类型
* `string`：字段的值必须是字符串类型
* `min`：字段的值必须大于等于指定值
* `max`：字段的值必须小于等于指定值
* `range`：字段的值长度必须指定范围内，包含 min 和 max。
* `regex`：字段的值必须匹配指定的正则表达式
* `in`：字段的值必须在指定的数组内
* `ip`：字段的值必须是一个合法的 ip
* `dateFormat`：字段的值必须是一个合法的 date format 格式。参考：http://php.net/manual/en/datetime.createfromformat.php
* `dateBefore`：字段的值必须小于给定时间。给定时间值非时间戳格式，而是 date/time 字符串，如：10 September 2000
* `dateAfter`：字段的值必须大于给定时间。时间值格式同上。
* `json`：字段的值必须是一个合法的 JSON 格式

## Note

#### 关于空字符串
如果传入一个数组:

```php
$data = [
    'foo' => '',
];
```

同时规则设置为:
```php
$rules= [
    'foo' => 'string',
];
```

此时校验结果会是通过的. 也就是说 `validator` 不会把 `''` 当作 `null` 处理.


## 参考
[https://github.com/vlucas/valitron](https://github.com/vlucas/valitron)

[https://github.com/illuminate/validation](https://github.com/illuminate/validation)


CHANGELOG.md
# Changelog For runner/validator

## [1.0.0] - 2018-12-25

### Added
- 增加可引入自定义语言包
- 增加自定义错误消息

### Changed
- 拆分 validate 及消息处理为 trait
- extend 的 callback 类作用域取消绑定到 Validator 示例. validate 函数参数增加传入 Validator 实例
- validate 从 protected 改为 publish

## [v1.0.1] - 2019-03-04
### Added
- 增加支持通过魔术方法调用扩展规则

tests/BaseRuleTest.php
<?php
/**
 * @author: RunnerLee
 * @email: runnerleer@gmail.com
 * @time: 17-2-24 17:40
 */
require __DIR__.'/ValidatorTester.php';

class BaseRuleTest extends \PHPUnit_Framework_TestCase
{
    protected $validator;

    public function setUp()
    {
        $this->validator = new ValidatorTester(['password' => '123456', 'password_confirm' => '123456'], []);
    }

    public function testSize()
    {
        $this->assertSame(true, $this->validator->callValidateRule('validateSize', '', 'a', 1));
        $this->assertSame(false, $this->validator->callValidateRule('validateSize', '', 'a', 2));
    }

//
//
//    public function testAccept()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateAccept', 'yes'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateAccept', 'no'));
//    }
//
//    public function testNumeric()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateNumeric', '1'));
//        $this->assertSame(true, $this->validator->callValidateRule('validateNumeric', '1.1'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateNumeric', 'runnerlee'));
//    }
//
//    public function testInteger()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateInteger', '1'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateInteger', '1.1'));
//    }
//
//    public function testFloat()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateFloat', '1.1'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateFloat', '1.1.1'));
//    }
//
//    public function testUrl()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateUrl', 'https://github.com'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateUrl', 'github.com'));
//    }
//
//    public function testBoolean()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateBoolean', '1'));
//        $this->assertSame(true, $this->validator->callValidateRule('validateBoolean', '0'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateBoolean', 'runnerlee'));
//    }
//
//    public function testConfirm()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateConfirm', '123456', 'password_confirm'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateConfirm', '12345', 'password_confirm'));
//    }
//
//    public function testRequired()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateRequired', 'runnerlee'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateRequired', null));
//    }
//
//    public function testDate()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateDate', '1995-09-06'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateDate', 'runnerlee'));
//    }
//
//    public function testEmail()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateEmail', 'runnerleer@gmail.com'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateEmail', 'runnerleer@gmail'));
//    }
//
//    public function testArray()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateArray', []));
//        $this->assertSame(false, $this->validator->callValidateRule('validateArray', 'runnerleer@gmail.com'));
//    }
//
//    public function testNullable()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateNullable', null));
//        $this->assertSame(true, $this->validator->callValidateRule('validateNullable', 'runnerlee'));
//    }
//
//    public function testMin()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateMin', 'string', 6));
//        $this->assertSame(true, $this->validator->callValidateRule('validateMin', '6', 6));
//        $this->assertSame(true, $this->validator->callValidateRule('validateMin', range(0, 5), 6));
//        $this->assertSame(false, $this->validator->callValidateRule('validateMin', 'string', 7));
//        $this->assertSame(false, $this->validator->callValidateRule('validateMin', '6', 7));
//        $this->assertSame(false, $this->validator->callValidateRule('validateMin', range(0, 5), 7));
//    }
//
//    public function testMax()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateMax', 'string', 6));
//        $this->assertSame(true, $this->validator->callValidateRule('validateMax', '6', 6));
//        $this->assertSame(true, $this->validator->callValidateRule('validateMax', range(0, 5), 6));
//        $this->assertSame(false, $this->validator->callValidateRule('validateMax', 'string', 5));
//        $this->assertSame(false, $this->validator->callValidateRule('validateMax', '6', 5));
//        $this->assertSame(false, $this->validator->callValidateRule('validateMax', range(0, 5), 5));
//    }
//
//    public function testRange()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateRange', 6, 6));
//        $this->assertSame(true, $this->validator->callValidateRule('validateRange', 6, 6, ''));
//        $this->assertSame(true, $this->validator->callValidateRule('validateRange', 6, 6, 7));
//        $this->assertSame(true, $this->validator->callValidateRule('validateRange', 6, '', 7));
//        $this->assertSame(false, $this->validator->callValidateRule('validateRange', 6));
//        $this->assertSame(false, $this->validator->callValidateRule('validateRange', 6, 7));
//        $this->assertSame(false, $this->validator->callValidateRule('validateRange', 6, '', 5));
//        $this->assertSame(false, $this->validator->callValidateRule('validateRange', 6, 3, 4));
//        $this->assertSame(false, $this->validator->callValidateRule('validateRange', 6, '', 4));
//    }
//
//    public function testRegex()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateRegex', '123456', '\d+'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateRegex', '123456', '\s+'));
//    }
//
//    public function testIn()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateIn', 'hello', 'hello', 'world'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateIn', 'hack', 'hello', 'world'));
//    }
//
//    public function testIp()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateIp', '114.114.114.114'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateIp', '114.114.114'));
//    }
//
//    public function testDateFormat()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateDateFormat', '1995-09-06', 'Y-m-d'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateDateFormat', '1995-09-06', 'H:i:s'));
//    }
//
//    public function testJson()
//    {
//        $this->assertSame(true, $this->validator->callValidateRule('validateJson', '{}'));
//        $this->assertSame(false, $this->validator->callValidateRule('validateJson', 'runnerlee'));
//    }
}


tests/ValidateTest.php
<?php

use PHPUnit\Framework\TestCase;
use Runner\Validator\Validator;

class ValidateTest extends TestCase
{
    public function testSize()
    {
        $validator = new Validator([
            'str1' => 'a',
            'arr1' => [1],
            'int1' => 1,
        ], [
            'str1' => 'size:1',
            'arr1' => 'size:1',
            'int1' => 'size:1',
        ]);

        $this->assertTrue($validator->validate());
    }
}


tests/ValidatorTest.php
<?php
/**
 * @author: RunnerLee
 * @email: runnerleer@gmail.com
 * @time: 17-2-21 11:13
 */
use Runner\Validator\Validator;

class ValidatorTest extends \PHPUnit_Framework_TestCase
{
    public function testValidator()
    {
        Validator::addExtension('channel_range', function ($field, $value, array $parameters, $validator) {
            $this->assertArrayHasKey($field, $validator->data());

            return false !== array_search($value, ['google', 'bing']);
        });

        $validator = new Validator($this->getData(), $this->getRules());

        $this->assertSame(true, $validator->validate());
        $this->assertSame($this->getData(), $validator->data());

        $data = [
            'range_a'       => '1.4',
            'range_b'       => '1.4',
            'channel_range' => 'baidu',
        ];

        $validator = new Validator(
            $data,
            [
                'do_not_has_field' => 'required|size:1',
                'range_a'          => 'range',
                'range_b'          => 'range:,',
                'channel_range'    => 'channel_range',
            ]
        );

        $this->assertSame(false, $validator->validate());
        $this->assertSame($data, $validator->data());
        $this->assertSame([
            'do_not_has_field',
            'range_a',
            'range_b',
            'channel_range',
        ], $validator->fails());
    }

    public function testMessage()
    {
        $data = [
            'data' => 'demo',
        ];
        $rules = [
            'data' => 'size:3',
        ];
        $validator = new Validator($data, $rules);

        $this->assertSame(false, $validator->validate());
        $this->assertSame("data 's size must be 3", $validator->messages()['data']['Size']);

        Validator::addExtension(
            'testing',
            function () {
                return false;
            },
            false,
            'is testing'
        );

        $validator = new Validator(
            [
                'a' => 'b',
            ],
            [
                'a' => 'testing',
            ]
        );

        $this->assertSame(false, $validator->validate());
        $this->assertSame(
            [
                'a' => [
                    'Testing' => 'a is testing',
                ],
            ],
            $validator->messages()
        );
    }

    public function testAddExtension()
    {
        Validator::addExtension('runner', function () {
            return true;
        });

        $data = [
            'a' => 'string',
        ];

        $validator = new Validator($data, [
            'a' => 'runner'
        ]);

        $this->assertSame(true, $validator->validate());

        $this->assertSame(true, $validator->validateRunner('a', 'a', []));

        $validator = new Validator($data, [
            'a' => 'demo',
        ]);

        $this->expectException(BadMethodCallException::class);
        $validator->validate();
    }

    public function testReplaceMessage()
    {
        $validator = new Validator(
            [
                'runnerlee' => 'hhh',
            ],
            [
                'runnerlee' => 'integer',
            ],
            [
                'integer' => 'is the most handsome man',
            ]
        );
        $validator->validate();

        $this->assertSame(
            [
                'runnerlee' => [
                    'Integer' => 'runnerlee is the most handsome man',
                ],
            ],
            $validator->messages()
        );
    }

    protected function getRules()
    {
        return [
            'data'           => 'size:4',
            'url'            => 'url',
            'array'          => 'required|array|min:1',
            'array.hello'    => 'numeric',
            'accept'         => 'accept',
            'integer'        => 'integer',
            'boolean'        => 'boolean',
            'string'         => 'string',
            'confirm'        => 'confirm:string',
            'date'           => 'date',
            'email'          => 'email',
            'numeric_string' => 'string|size:6',
            'block_reason'   => 'string|required_with:is_blocked',
            'channel'        => 'required|channel_range',
            'diff_beta'      => 'diff:diff_alpha',
            'json'           => 'json',
            'ip'             => 'ip',
            'in'             => 'in:a,b,c',
            'float'          => 'required_unless:channel,baidu|float|size:1.44|min:1.44|max:1.44',
            'sub_arr.demo'   => 'size:3',
            'without_alpha'  => 'required_without:without_beta|required_if:channel,google',
            'nullable'       => 'nullable',
            'range_a'        => 'range:1,1.5',
            'range_b'        => 'range:,1.5',
            'range_c'        => 'range:1,',
            'range_g'        => 'range:1',
            'regex'          => 'regex:#\w+#',
            'date_format'    => 'date_format:Y-m-d',
            'date_before'    => 'date_before:2018-01-02',
            'date_after'     => 'date_after:2017-12-31',
        ];
    }

    protected function getData()
    {
        return [
            'data'  => 'demo',
            'url'   => 'http://runnerlee.com',
            'array' => [
                'hello' => '1.1',
            ],
            'accept'         => 'on',
            'integer'        => '1',
            'boolean'        => 0,
            'string'         => 'hello world',
            'confirm'        => 'hello world',
            'date'           => '1995-09-06',
            'email'          => 'runnerleer@gmail.com',
            'numeric_string' => '123456',
            'is_blocked'     => 'no',
            'block_reason'   => '123',
            'channel'        => 'google',
            'diff_alpha'     => 'hello',
            'diff_beta'      => 'world',
            'json'           => '{"a":"b"}',
            'ip'             => '127.0.0.1',
            'in'             => 'a',
            'float'          => '1.44',
            'sub_arr'        => [],
            'without_alpha'  => 'a',
            'range_a'        => '1.5',
            'range_b'        => '1.5',
            'range_c'        => '1.5',
            'range_e'        => '1.5',
            'range_f'        => '1.5',
            'range_g'        => '3',
            'regex'          => '123abc',
            'date_format'    => '2018-01-01',
            'date_before'    => '2018-01-01',
            'date_after'     => '2018-01-01',
            'nullable'       => null,
        ];
    }
}


src/message.php
<?php
/**
 * @author: RunnerLee
 * @email: runnerleer@gmail.com
 * @time: 17-2-27 10:15
 */
return [
    'Required'   => 'is required',
    'Accepted'   => 'must be accepted',
    'Numeric'    => 'must be numeric',
    'Integer'    => 'must be an integer',
    'Float'      => 'must be a float',
    'Size'       => "'s size must be %s",
    'Min'        => 'must be at least %d',
    'Max'        => 'must be no more than %d',
    'In'         => 'contains invalid value',
    'Ip'         => 'is not a valid IP address',
    'Email'      => 'is not a valid email address',
    'Url'        => 'is not a valid URL',
    'Regex'      => 'contains invalid characters',
    'Date'       => 'is not a valid date',
    'DateFormat' => "must be date with format '%s'",
    'Boolean'    => 'must be a boolean',
    'Range'      => 'must be between %d and %d characters',
    'Confirm'    => 'must be same as %s field',
    'Array'      => 'must be an array',
    'String'     => 'must be a string',
    'Json'       => 'must be a json string',
];


src/Validator.php
<?php
/**
 * @author: RunnerLee
 * @email: runnerleer@gmail.com
 * @time: 17-2-20 15:22
 */

namespace Runner\Validator;

use BadMethodCallException;
use Runner\Validator\Concerns\MessagesAttributes;
use Runner\Validator\Concerns\ValidatesAttributes;

/**
 * Class Validator.
 */
class Validator
{
    use ValidatesAttributes, MessagesAttributes;

    /**
     * @var array
     */
    protected $data = [];

    /**
     * @var array
     */
    protected $ruleGroups = [];

    /**
     * @var array
     */
    protected $messages = [];

    /**
     * @var array
     */
    protected static $forceRules = ['Required', 'RequiredIf', 'RequiredWith', 'RequiredUnless', 'RequiredWithout'];

    /**
     * @var array
     */
    protected static $extensions = [];

    /**
     * @var array
     */
    protected static $extensionTemplates = [];

    /**
     * Validator constructor.
     *
     * @param array  $data
     * @param array  $rules
     * @param array  $customMessages
     * @param string $file
     */
    public function __construct(array $data, array $rules, array $customMessages = [], $file = __DIR__.'/messages/en.php')
    {
        $this->data = $data;
        $this->parseRules($rules);
        $this->loadMessageTemplate($file, $customMessages);
    }

    /**
     * @param $name
     * @param $callback
     * @param bool   $isForce
     * @param string $message
     */
    public static function addExtension($name, $callback, $isForce = false, $message = null)
    {
        $name = self::formatRuleName($name);

        self::$extensions[$name] = $callback;

        $isForce && self::$forceRules[] = $name;

        !empty($message) && (static::$extensionTemplates[$name] = $message);
    }

    /**
     * @return bool
     */
    public function validate()
    {
        foreach ($this->ruleGroups as $field => $rules) {
            if ($this->hasField($field)) {
                $value = $this->getField($field);
                foreach ($rules as $rule => $parameters) {
                    if (!$this->runValidateRule($rule, $field, $value, $parameters)) {
                        $this->messages[$field][$rule] = $this->buildMessage($rule, $field, $parameters);
                    }
                }
            } elseif ($forceRules = array_intersect(self::$forceRules, array_keys($rules))) {
                foreach ($forceRules as $rule) {
                    if (!$this->runValidateRule($rule, $field, null, $rules[$rule])) {
                        $this->messages[$field][$rule] = $this->buildMessage($rule, $field, $rules[$rule]);
                    }
                }
            }
        }

        return 0 === count($this->messages);
    }

    /**
     * @return array
     */
    public function fails()
    {
        return array_keys($this->messages);
    }

    /**
     * @return array
     */
    public function messages()
    {
        return $this->messages;
    }

    /**
     * @return array
     */
    public function data()
    {
        return $this->data;
    }

    /**
     * @param array $ruleGroups
     */
    protected function parseRules(array $ruleGroups)
    {
        $map = [];
        foreach ($ruleGroups as $field => $rules) {
            foreach (explode('|', $rules) as $rule) {
                list($rule, $parameters) = explode(':', (false === strpos($rule, ':') ? ($rule.':') : $rule), 2);
                !isset($map[$rule]) && $map[$rule] = self::formatRuleName($rule);
                $rule = $map[$rule];
                $this->ruleGroups[$field][$rule] = ('' === $parameters ? [] : explode(',', $parameters));
            }
        }
    }

    /**
     * @param $name
     *
     * @return string
     */
    protected static function formatRuleName($name)
    {
        return implode(
            '',
            array_map(
                function ($value) {
                    return ucfirst($value);
                },
                explode('_', $name)
            )
        );
    }

    /**
     * @param string $field
     *
     * @return bool
     */
    protected function hasField($field)
    {
        $field = explode('.', $field);
        $item = array_shift($field);
        if (!array_key_exists($item, $this->data)) {
            return false;
        }
        $value = $this->data[$item];

        foreach ($field as $item) {
            if (!array_key_exists($item, $value)) {
                return false;
            }
            $value = $value[$item];
        }

        return true;
    }

    /**
     * @param string $field
     *
     * @return mixed
     */
    protected function getField($field)
    {
        $field = explode('.', $field);
        $item = array_shift($field);
        $value = $this->data[$item];
        foreach ($field as $item) {
            $value = $value[$item];
        }

        return $value;
    }

    /**
     * @param $field
     * @param $value
     * @param $rule
     * @param array $parameters
     *
     * @return bool
     */
    protected function runValidateRule($rule, $field, $value, array $parameters = [])
    {
        $callback = array_key_exists($rule, self::$extensions) ? self::$extensions[$rule] : [$this, "validate{$rule}"];

        return (bool) call_user_func($callback, $field, $value, $parameters, $this);
    }

    public function __call($method, $arguments)
    {
        $rule = self::formatRuleName(substr($method, 8));

        if (!isset(self::$extensions[$rule])) {
            throw new BadMethodCallException(sprintf('Method %s::%s does not exists', static::class, $method));
        }

        return $this->runValidateRule($rule, ...$arguments);
    }
}


src/Concerns/MessagesAttributes.php
<?php
/**
 * @author: RunnerLee
 * @email: runnerleer@gmail.com
 * @time: 2018-12
 */

namespace Runner\Validator\Concerns;

use Exception;

trait MessagesAttributes
{
    protected $templates = [];

    protected function loadMessageTemplate($file, array $customMessage = [])
    {
        $this->templates = require $file;
        $this->templates = array_merge($this->templates, self::$extensionTemplates);

        foreach ($customMessage as $rule => $template) {
            $this->templates[self::formatRuleName($rule)] = $template;
        }
    }

    /**
     * @param string $rule
     * @param string $field
     * @param array  $parameters
     *
     * @return string
     */
    protected function buildMessage($rule, $field, array $parameters = [])
    {
        if (!isset($this->templates[$rule])) {
            return "{$field} field check failed";
        }
        array_unshift($parameters, "{$field} {$this->templates[$rule]}");

        try {
            return sprintf(...$parameters);
        } catch (Exception $e) {
            return "{$field} filed check failed";
        }
    }
}


src/messages/en.php
<?php
/**
 * @author: RunnerLee
 * @email: runnerleer@gmail.com
 * @time: 17-2-27 10:15
 */

return [
    'Required'   => 'is required',
    'Accepted'   => 'must be accepted',
    'Numeric'    => 'must be numeric',
    'Integer'    => 'must be an integer',
    'Float'      => 'must be a float',
    'Size'       => "'s size must be %s",
    'Min'        => 'must be at least %d',
    'Max'        => 'must be no more than %d',
    'In'         => 'contains invalid value',
    'Ip'         => 'is not a valid IP address',
    'Email'      => 'is not a valid email address',
    'Url'        => 'is not a valid URL',
    'Regex'      => 'contains invalid characters',
    'Date'       => 'is not a valid date',
    'DateFormat' => "must be date with format '%s'",
    'Boolean'    => 'must be a boolean',
    'Range'      => 'must be between %d and %d characters',
    'Confirm'    => 'must be same as %s field',
    'Array'      => 'must be an array',
    'String'     => 'must be a string',
    'Json'       => 'must be a json string',
];
